README
# Acow

Acow is a command line accounting application that I wrote as an experiment in the use of SQL (using MySQL in particular).

It is a relatively simple system, but supports multiple accounts and transactions between accounts.
Note that the application is entirely console based, however, it could be combined with a graphical frontend.

The perl directory contains the first version, written in perl using the DBI database module to connect to MySQL.
The c directory contains a C version, which uses plain text files in place of a database to store the account details.


ChangeLog

2007-04-20 Andrew Wedgbury <wedge@sconemad.com>

	* rewrite in c.
	
2006-05-03 Andrew Wedgbury <wedge@sconemad.com>

	* add copy command.

2006-03-09 Andrew Wedgbury <wedge@sconemad.com>

        * add multi-command interactive mode.
	
2006-03-02 Andrew Wedgbury <wedge@sconemad.com>

	* commit initial working version (using Perl and MySQL).
		
2006-02-22 Andrew Wedgbury <wedge@sconemad.com>

	* started.


c/acow.c
/* acow - simple accounting program (http://www.sconemad.com)

acow_file - File operations

Copyright (c) 2005-2007 Andrew Wedgbury <wedge@sconemad.com>

This program is free software; you can redistribute it and/or modify
it under the terms of the GNU General Public License as published by
the Free Software Foundation; either version 2, or (at your option)
any later version.

This program is distributed in the hope that it will be useful,
but WITHOUT ANY WARRANTY; without even the implied warranty of
MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
GNU General Public License for more details.

You should have received a copy of the GNU General Public License
along with this program (see the file COPYING); if not, write to the
Free Software Foundation, Inc.,
59 Temple Place - Suite 330, Boston, MA  02111-1307, USA */

#include "acow.h"
#include <getopt.h>

const char filename[] = "test.acow";

/* ========================================================================= */
int list_account_callback(
  ACOW_TRANSACTION* tra,
  enum AcowTransactionType type,
  int balance,
  void* user_data
)
{
  ACOW_DATA* data = (ACOW_DATA*)user_data;
  ACOW_ACCOUNT* acc =0;
  struct tm tmstruct;
  char tmstring[32];

  /* Convert time to a string */
  localtime_r(&tra->timedate,&tmstruct);
  strftime(tmstring,31,"%Y-%m-%d",&tmstruct);
      
  switch (type) {
    case Checkpoint:
      acc = lookup_account_id(data,tra->dest);
      printf("%6u %s  =BALANCE CHECKPOINT=                               %9.2f=\n",
             tra->id,tmstring,
             balance*0.01);
      break;

    case Debit:
      acc = lookup_account_id(data,tra->dest);
      printf("%6u %s  %-30s %9.2f           %9.2f\n",
             tra->id,tmstring,acc->name,
             tra->value*0.01,balance*0.01);
      break;
      
    case Credit:
      acc = lookup_account_id(data,tra->src);
      printf("%6u %s  %-30s           %9.2f %9.2f\n",
             tra->id,tmstring,acc->name,
             tra->value*0.01,balance*0.01);
      break;
  }
}

/* ========================================================================= */
int list_account(ACOW_DATA* data, unsigned int acc_id)
{
  int i,num=0;
  ACOW_TRANSACTION* tra;
  ACOW_ACCOUNT* acc;
  int balance=0,debit_tot=0,credit_tot=0;

  acc = lookup_account_id(data,acc_id);
  printf("Listing transactions for account: %s (ID=%u)\n\n",
         acc->name,acc_id);

  printf("    ID    DATE     PAYEE                              DEBIT    CREDIT   BALANCE\n");
  printf("--------------------------------------------------------------------------------\n");

  enumerate_transactions(data,acc_id,0,0,list_account_callback,data);
 
  printf("--------------------------------------------------------------------------------\n");
}

/* ========================================================================= */
int list_all_accounts(ACOW_DATA* data)
{
  int i;
  ACOW_ACCOUNT* acc;

  printf("    ID  NAME                             BALANCE\n");
  printf("--------------------------------------------------------------------------------\n");

  for (i=0; i < data->num_accounts; ++i) {
    acc = data->accounts + i;
    if (0 != acc->user) {
      printf( "%6u  %-30s %9.2f\n",
              acc->id,
              acc->name,
              account_balance(data,acc->id)*0.01 );
    }
  }

  return 0;
}

/* ========================================================================= */
time_t parse_date(const char* str)
{
  struct tm tmstruct;
  memset(&tmstruct,0,sizeof(struct tm));

  if (3 != sscanf(str,"%d-%d-%d",
                  &tmstruct.tm_year,&tmstruct.tm_mon,&tmstruct.tm_mday)) {
    return -1;
  }

  /* Month should be zero-based */
  --tmstruct.tm_mon;

  /* Years start at 1900 */
  tmstruct.tm_year -= 1900;

  /* Turn it into a time_t */
  return mktime(&tmstruct);
}

/* ========================================================================= */
int main(int argc,char* argv[])
{
  int c=0;
  int err=0;
  ACOW_DATA data;

  /* options */
  extern char* optarg;
  ACOW_ACCOUNT* acc=0;
  ACOW_ACCOUNT* pay=0;
  time_t datetime=0;
  int value=0;
  unsigned int id=0;

  memset(&data,0,sizeof(ACOW_DATA));
  data.accounts = 0;
  data.transactions = 0;

  char* cmd = argv[1];
  if (!cmd) exit(0);

  load_file(&data,filename);

  while ((c = getopt(argc,argv,"?a:d:t:p:v:c:f:i:e")) >= 0) {
    switch (c) {
      case '?': 
        printf("HELP\n"); 
        exit(1);
        break;

      case 'a': // Account 
        acc = lookup_account_name(&data,optarg); 
        if (acc == 0) {
          printf("ERROR: Unknown account '%s'\n",optarg);
          exit(1);
        }
        break;

      case 'd': // Date
        datetime = parse_date(optarg);
        if (datetime < 0) {
          printf("ERROR: Invalid date '%s'\n",optarg);
          exit(1);
        }
        break;

      case 't': // Type
        break;

      case 'p': // Payee
        pay = lookup_account_name(&data,optarg); 
        if (pay == 0) {
          printf("ERROR: Unknown payee '%s'\n",optarg);
          exit(1);
        }
        break;

      case 'v': // Value
        if (1 != sscanf(optarg,"%d",&value)) {
          printf("ERROR: Invalid transaction value '%s'\n",optarg);
          exit(1);
        }
        break;

      case 'c': // Comment
        break;
      case 'f': // File
        break;

      case 'i': // ID
        if (1 != sscanf(optarg,"%u",&id)) {
          printf("ERROR: Invalid transaction id '%s'\n",optarg);
          exit(1);
        }
        break;
      case 'e': // Extended
        break;
    }
  }

  if (0 == strcmp(cmd,"ls") ||
      0 == strcmp(cmd,"list")) {
    if (acc == 0) {
      printf("ERROR: No account specified\n");
      exit(1);
    }
    list_account(&data,acc->id);

  } else if (0 == strcmp(cmd,"acc")) {
    list_all_accounts(&data);

  } else {
    printf("ERROR: Unknown command '%s'\n",cmd);
    exit(1);
  }

  save_file(&data,filename);

  exit(0);
}


c/acow.h
/* acow - simple accounting program (http://www.sconemad.com)

Global definitions

Copyright (c) 2005-2007 Andrew Wedgbury <wedge@sconemad.com>

This program is free software; you can redistribute it and/or modify
it under the terms of the GNU General Public License as published by
the Free Software Foundation; either version 2, or (at your option)
any later version.

This program is distributed in the hope that it will be useful,
but WITHOUT ANY WARRANTY; without even the implied warranty of
MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
GNU General Public License for more details.

You should have received a copy of the GNU General Public License
along with this program (see the file COPYING); if not, write to the
Free Software Foundation, Inc.,
59 Temple Place - Suite 330, Boston, MA  02111-1307, USA */

#include <stdlib.h>
#include <stdio.h>
#include <string.h>
#include <errno.h>
#include <time.h>

typedef struct {
  unsigned int id;   // Unique ID
  time_t timedate;   // Time/date of transaction
  unsigned int src;  // Source account ID
  unsigned int dest; // Destination account ID
  int value;         // Transaction value
  char* comment;     // Space for comment if required
} ACOW_TRANSACTION;

typedef struct {
  unsigned int id;   // Unique ID
  unsigned int user; // User ID (0 for external account)
  char* name;        // Account name
  char* comment;     // Space for comment if required
} ACOW_ACCOUNT;

typedef struct {
  unsigned int num_accounts;
  unsigned int max_accounts;
  ACOW_ACCOUNT* accounts;
  unsigned int num_transactions;
  unsigned int max_transactions;
  ACOW_TRANSACTION* transactions;
} ACOW_DATA;

enum AcowTransactionType {
  Credit,
  Debit,
  Checkpoint
};

/* ========================================================================= */
/* acow_calc */

/* Enumerate transactions in a given account */
int enumerate_transactions(
  ACOW_DATA* data,
  unsigned int acc_id,
  time_t start,
  time_t end,
  int (*transaction_callback)(
    ACOW_TRANSACTION* tra,
    enum AcowTransactionType type,
    int balance,
    void* user_data),
  void* user_data
);

/* Obtain the current balance for a given account */
int account_balance(
  ACOW_DATA* data,
  unsigned int acc_id
);


/* ========================================================================= */
/* acow_data */

/* Resize data arrays as required */
int realloc_data(
  ACOW_DATA* data,
  int num_accounts,
  int num_transactions
);

/* Re-sort the data arrays */
int resort_data(
  ACOW_DATA* data
);

/* Lookup account by id or name */
ACOW_ACCOUNT* lookup_account_id(ACOW_DATA* data, unsigned int id);
ACOW_ACCOUNT* lookup_account_name(ACOW_DATA* data, const char* name);

/* Add an account */
int add_account(
  ACOW_DATA* data,
  unsigned int user,
  char* name,
  char* comment
);

/* Remove an account */
int remove_account(
  ACOW_DATA* data,
  unsigned int id
);

/* Lookup transaction by id */
ACOW_TRANSACTION* lookup_transaction_id(ACOW_DATA* data, unsigned int id);

/* Add a transaction */
int add_transaction(
  ACOW_DATA* data,
  time_t timedate,
  unsigned int src,
  unsigned int dest,
  int value,
  char* comment
);

/* Remove a transaction */
int remove_transaction(
  ACOW_DATA* data,
  unsigned int id
);

/* ========================================================================= */
/* acow_file */

/* Load from a file */
int load_file(
  ACOW_DATA* data,
  const char* path
);

/* Save to a file */
int save_file(
  ACOW_DATA* data,
  const char* path
);  


c/utest.c
/* acow - simple accounting program (http://www.sconemad.com)

utest - Acow unit tests

Copyright (c) 2005-2007 Andrew Wedgbury <wedge@sconemad.com>

This program is free software; you can redistribute it and/or modify
it under the terms of the GNU General Public License as published by
the Free Software Foundation; either version 2, or (at your option)
any later version.

This program is distributed in the hope that it will be useful,
but WITHOUT ANY WARRANTY; without even the implied warranty of
MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
GNU General Public License for more details.

You should have received a copy of the GNU General Public License
along with this program (see the file COPYING); if not, write to the
Free Software Foundation, Inc.,
59 Temple Place - Suite 330, Boston, MA  02111-1307, USA */

#include "acow.h"

const char filename[] = "test.acow";

#define UTEST(expected,cmd) \
  printf("UTEST: running " #cmd "\n"); \
  err = cmd; \
  if (err != expected) {\
    printf("\n### TEST FAILED got %d, expected %d, errno %d ###\n", \
           err,expected,errno); \
    exit(1); \
  }

/* ========================================================================= */
int main(int argc,char* argv[])
{

  int err;
  ACOW_DATA data;
  ACOW_ACCOUNT* acc;
  memset(&data,0,sizeof(ACOW_DATA));
  data.accounts = 0;
  data.transactions = 0;

  //  load_file(&data,filename);
  
  UTEST(1,add_account(&data,1,"current","My current account"));
  UTEST(2,add_account(&data,1,"savings","My savings account"));

  UTEST(3,add_account(&data,0,"wizzycorp","Generic company/employer"));
  UTEST(4,add_account(&data,0,"bank loan","A loan from the bank"));

  UTEST(5,add_account(&data,0,"ultra markt","Generic store 1"));
  UTEST(6,add_account(&data,0,"price busterz","Generic store 2"));
  UTEST(7,add_account(&data,0,"expensives r us","Generic store 3"));

  UTEST(1,add_transaction(&data,0,2,2,4321,"Initial savings balance"));
  UTEST(2,add_transaction(&data,0,4,1,9999,"Take out bank loan"));

  UTEST(3,add_transaction(&data,0,1,2,123,"Save one twenty three"));

  UTEST(4,add_transaction(&data,0,1,5,1,"Spend a penny"));
  UTEST(5,add_transaction(&data,0,1,6,100,"Spend a pound"));
  UTEST(6,add_transaction(&data,0,1,7,10000,"Spend lots"));

  UTEST(7,add_transaction(&data,0,2,1,20000,"Plunder savings"));

  UTEST(0,remove_transaction(&data,4));
  UTEST(-1,remove_transaction(&data,4));

  save_file(&data,"test.acow");

  printf("\nTEST PASSED!\n");
  exit(0);
}


c/acow_calc.c
/* acow - simple accounting program (http://www.sconemad.com)

acow_calc - Account calculations

Copyright (c) 2005-2007 Andrew Wedgbury <wedge@sconemad.com>

This program is free software; you can redistribute it and/or modify
it under the terms of the GNU General Public License as published by
the Free Software Foundation; either version 2, or (at your option)
any later version.

This program is distributed in the hope that it will be useful,
but WITHOUT ANY WARRANTY; without even the implied warranty of
MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
GNU General Public License for more details.

You should have received a copy of the GNU General Public License
along with this program (see the file COPYING); if not, write to the
Free Software Foundation, Inc.,
59 Temple Place - Suite 330, Boston, MA  02111-1307, USA */

#include "acow.h"
#include <limits.h>

/* ========================================================================= */
int enumerate_transactions(
  ACOW_DATA* data,
  unsigned int acc_id,
  time_t start,
  time_t end,
  int (*transaction_callback)(
    ACOW_TRANSACTION*,
    enum AcowTransactionType,
    int,
    void*),
  void* user_data
)
{
  int i,num=0;
  ACOW_TRANSACTION* tra;
  int balance=0;
  enum AcowTransactionType type;

  if (end == 0) end = INT_MAX;

  for (i=0; i < data->num_transactions; ++i) {
    tra = data->transactions + i;
    if (tra->timedate > end) break;

    if (tra->src == acc_id || tra->dest == acc_id) {
      /* This transaction affects this account*/
      if (tra->src == acc_id && tra->dest == acc_id) {
        balance = tra->value; /* Checkpoint */
        type = Checkpoint;
      } else if (tra->src == acc_id) {
        balance -= tra->value; /* Debit */
        type = Debit;
      } else if (tra->dest == acc_id) {
        balance += tra->value; /* Credit */
        type = Credit;
      }

      // Call user-supplied function if the transaction is
      // within the required date range
      if (tra->timedate >= start) {
        ++num;
        transaction_callback(tra,type,balance,user_data);
      }
    }
  }

  return num;
}

/* ========================================================================= */
int account_balance_tra_callback(
  ACOW_TRANSACTION* tra,
  enum AcowTransactionType type,
  int balance,
  void* user_data
)
{
  *((int*)user_data) = balance;
}

/* ========================================================================= */
int account_balance(
  ACOW_DATA* data,
  unsigned int acc_id
)
{
  int balance =0;
  enumerate_transactions(data,acc_id,0,0,account_balance_tra_callback,&balance);
  return balance;
}



perl/acow.pl
#!/usr/bin/perl -w
#
# acow - simple accounting (http://www.sconemad.com/acow)
#
# Copyright (c) 2005-2007 Andrew Wedgbury <wedge@sconemad.com>
#
# This program is free software; you can redistribute it and/or modify
# it under the terms of the GNU General Public License as published by
# the Free Software Foundation; either version 2, or (at your option)
# any later version.
#
# This program is distributed in the hope that it will be useful,
# but WITHOUT ANY WARRANTY; without even the implied warranty of
# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
# GNU General Public License for more details.
#
# You should have received a copy of the GNU General Public License
# along with this program (see the file COPYING); if not, write to the
# Free Software Foundation, Inc.,
# 59 Temple Place - Suite 330, Boston, MA  02111-1307, USA

use DBI;
use Getopt::Std;

my $ACOW_VER = "0.1.0";
my $RCFILE = $ENV{HOME}.'/.acowrc';

my $compress = 0;
my $MFMT = "%9.02f";
my $TRANS_FMT = "%6d %s  %-30s %9s %9s %9s\n";
my $ACC_FMT = " %-32s %5d $MFMT\n";

my $LINE = ('-' x 80)."\n";
my $MIN_PAYEE = 1000;
my $PROMPT = "\nacow> ";

#===================================================================[ START ]==
# Check for simple standard info options
exit do_help() if (@ARGV==1 && $ARGV[0] =~ /^\-*(h|\?)/);
exit do_ver() if (@ARGV==1 && $ARGV[0] =~ /^\-*ver/);

# Set defaults then read settings from rcfile
my %rcopts;
$rcopts{dbuser} = $ENV{USER};
$rcopts{dbpasswd} = '';
$rcopts{dbhost} = 'localhost';
read_rcfile();
$rcopts{dbname} = 'acow_'.$rcopts{dbuser} if (!$rcopts{dbname});

# Connect to database
my $db = 'dbi:mysql:dbname='.$rcopts{dbname}.';host='.$rcopts{dbhost};
my $dbh = DBI->connect($db,$rcopts{dbuser},
		       $rcopts{dbpasswd},
		       { RaiseError => 1, AutoCommit => 0 });
die "Cannot connect to database" if (!$dbh);
do_init_tables();

my %opts;
my $account = 'current';
my $payee;
my $date;
my $type = 'd';
my $value;
my $comment = '';

if (@ARGV) {
    # Single command
    do_cmd();

} else {
    # Interactive multiple commands
    print $PROMPT;
    while (<>) {
	@ARGV = ();
	foreach my $arg (split(/(\"[^\"]*\")/,$_)) {
	    if ($arg =~ /\"([^\"]*)\"/) {
		push(@ARGV,$1);
	    } else {
		push(@ARGV,split(' ',$arg));
	    }
	}
	last if (do_cmd());
	print $PROMPT;
    }
}

$dbh->disconnect();


#----------------------------------------------------------
sub do_cmd
{
    if (scalar @ARGV == 0) {
	return 0;
    }
    my $cmd = '';
    $cmd = shift(@ARGV) if ($ARGV[0] !~ /^\-/);
    %opts = ();
    if (!getopts('a:d:t:p:v:c:f:i:e',\%opts)) {
	return 0;
    }

    # Get options or defaults
    $account = $opts{a} if ($opts{a});
    $payee = $opts{p};
    $date = $opts{d} if ($opts{d});
    $type = $opts{t} || 'd';
    $value = $opts{v};
    $comment = $opts{c} || '';
    
    # Commands
    if ($cmd eq 'help') {
	do_help();
    } elsif ($cmd =~ /^ver/) {
	do_ver();
    } elsif ($cmd eq 'add') {
	do_add();
    } elsif ($cmd =~ /^del/) {
	do_delete();
    } elsif ($cmd =~ /^edi/) {
	do_edit();
    } elsif ($cmd =~ /^cop/) {
	do_copy();
    } elsif ($cmd =~ /^lis/ || $cmd eq 'ls') {
	do_list();
    } elsif ($cmd =~ /^inf/) {
	do_info();
    } elsif ($cmd =~ /^acc/) {
	do_accounts(1);
    } elsif ($cmd =~ /^pay/) {
	do_accounts(0);
    } elsif ($cmd =~ /^imp/) {
	do_import();
    } elsif ($cmd eq 'compress') {
	$compress = 1;
	$MFMT = "%.2f";
	$TRANS_FMT = "%d,%s,%s,%s,%s,%s\n";
	$ACC_FMT = "%s,%d,$MFMT\n";
    } elsif ($cmd eq 'exit' || $cmd eq 'quit') {
	return 1;
    } else {
	print "ERROR: Unknown command\n";
    }
    return 0;
}

#----------------------------------------------------------
sub do_add
{
    # Date - default to today's date
    if (!defined $date) {
	my @t = localtime();
	$date = sprintf("%04d-%02d-%02d",1900+$t[5],1+$t[4],$t[3]);
    }
  
    # Payee
    if (!defined $payee && $type ne 'b') {
	print "ERROR: No payee account specified\n";
	return 1;
    }

    # Value
    if (!defined $value) {
	print "ERROR: No value specified\n";
	return 1;
    }
    
    my $account_id = lookup_account($account,0);
    if (!defined $account_id) {
	print "ERROR: Unknown account '$account'\n";
	return 1;
    }
    my $payee_id = $payee ? lookup_account($payee,1) : 0;
    my $src = ($type ne 'c') ? $account_id : $payee_id;
    my $dst = ($type ne 'd') ? $account_id : $payee_id;
    $value *= 100;

    $dbh->do(
      "INSERT INTO transaction (date, acc_src, acc_dst, value, comment) ".
      "VALUES (\"$date\", $src, $dst, $value, \"$comment\")");
}

#----------------------------------------------------------
sub do_delete
{
    my $id = $opts{i};
    if (!defined $id) {
	print "ERROR: Must specify transaction ID to delete\n";
	return 1;
    }
    my $r = $dbh->do(
      "DELETE FROM transaction WHERE id = $id");

    if ($r != 1) {
	print "ERROR: Unable to delete transaction $id\n";
	return 1;
    }
}

#----------------------------------------------------------
sub do_edit
{
    my $id = $opts{i};
    if (!defined $id) {
	print "ERROR: Must specify transaction ID to edit\n";
	return 1;
    }

    if (defined $opts{d}) {
	my $r = $dbh->do(
          "UPDATE transaction SET date = \"$date\" WHERE id = $id");
    }

    if (defined $opts{a} || defined $opts{p}) {

	my $r = $dbh->selectrow_arrayref(
          "SELECT acc_src, acc_dst FROM transaction WHERE id = $id");
	my ($src,$dst) = @{$r};

	my $account_id;
	if ($opts{a}) {
	    $account_id = lookup_account($account,0);
	    if (!defined $account_id) {
		print "ERROR: Unknown account '$account'\n";
		return 1;
	    }
	}
	my $payee_id;
	if ($opts{p}) {
	    $payee_id = lookup_account($payee,1);
	}
	if ($src < $MIN_PAYEE && defined $opts{a}) {
	    my $r = $dbh->do(
              "UPDATE transaction SET acc_src = $account_id WHERE id = $id");
	}
	if ($dst < $MIN_PAYEE && defined $opts{a}) {
	    my $r = $dbh->do(
              "UPDATE transaction SET acc_dst = $account_id WHERE id = $id");
	}
	if ($src > $MIN_PAYEE && defined $opts{p}) {
	    my $r = $dbh->do(
              "UPDATE transaction SET acc_src = $payee_id WHERE id = $id");
	}
	if ($dst > $MIN_PAYEE && defined $opts{p}) {
	    my $r = $dbh->do(
              "UPDATE transaction SET acc_dst = $payee_id WHERE id = $id");
	}
    }

    if (defined $opts{v}) {
	$value *= 100;
	my $r = $dbh->do(
          "UPDATE transaction SET value = $value WHERE id = $id");
    }

    if (defined $opts{c}) {
	my $r = $dbh->do(
          "UPDATE transaction SET comment = \"$comment\" WHERE id = $id");
    }

    if (defined $opts{t}) {
	print "ERROR: Cannot change transaction type\n";
	return 1;
    }
}

#----------------------------------------------------------
sub do_copy
{
    my $id = $opts{i};
    if (!defined $id) {
	print "ERROR: Must specify transaction ID to copy\n";
	return 1;
    }

    my $r = $dbh->selectrow_arrayref(
      "SELECT date, acc_src, acc_dst, value, comment ".
      "FROM transaction ".
      "WHERE id = '$id'");

    if (!$r) {
	print "ERROR: Transaction '$id' not found\n";
	return 1;
    }
    my ($c_date,$src,$dst,$c_value,$c_comment) = @{$r};
    $c_value /= 100;

    my $date = $opts{d} || $c_date;
    my $value = $opts{v} || $c_value;
    $value *= 100;
    my $comment = $opts{c} || $c_comment;

    $dbh->do(
      "INSERT INTO transaction (date, acc_src, acc_dst, value, comment) ".
      "VALUES (\"$date\", $src, $dst, $value, \"$comment\")");
}

#----------------------------------------------------------
sub do_list
{
    print "ACCOUNT: $account\n";
    print "\n";

    my $trans = get_transactions($account);

    my $final_bal = 0;
    my $credit_sum = 0;
    my $debit_sum = 0;

    if (!$compress) {
	print "    ID    DATE     PAYEE                             ".
	    " DEBIT    CREDIT   BALANCE\n";
	print $LINE;
    }

    foreach my $row (@{$trans}) {
	my ($id,$date,$payee,$comment,$debit,$credit,$bal) = @{$row};

	$credit_sum += $credit;
	$debit_sum += $debit;
	$credit_str = $credit ? sprintf($MFMT,$credit/100) : "";
	$debit_str = $debit ? sprintf($MFMT,$debit/100) : "";
	$bal_str = sprintf($MFMT,$bal/100);
	$final_bal = $bal;
	printf($TRANS_FMT,$id,$date,$payee,$debit_str,$credit_str,$bal_str);
	if ($comment && $opts{e}) {
	    print((" "x20)."($comment)\n");
	}
    }

    if (!$compress) {
	print $LINE;
	printf(" %-48s $MFMT $MFMT $MFMT\n",
	       "TOTALS FOR ".scalar @{$trans}. " TRANSACTIONS:", 
	       $debit_sum/100, $credit_sum/100, $final_bal/100);
	print $LINE;
    }
}

#----------------------------------------------------------
sub do_info
{
    my $id = $opts{i};
    if (!$id) {
	print "ERROR: Must specify transaction ID for info\n";
	return 1;
    }
    my $r = $dbh->selectrow_arrayref(
      "SELECT date, a1.name AS 'from', a2.name AS 'to', value, comment ".
      "FROM transaction AS t, account AS a1, account AS a2 ".
      "WHERE t.id = '$id' AND t.acc_src = a1.id AND t.acc_dst = a2.id");

    if (!$r) {
	print "ERROR: Transaction '$id' not found\n";
	return 1;
    }
    my ($date,$src,$dst,$value,$comment) = @{$r};
    $value /= 100;
    print "ID: $id\n".
	  "DATE: $date\n".
	  "SOURCE: $src\n".
	  "DEST: $dst\n".
	  "VALUE: $value\n".
	  "COMMENT: $comment\n";
}

#----------------------------------------------------------
sub do_accounts
{
    print "\n";
    my ($a) = @_;
    my $title = $a ? "ACCOUNT" : "PAYEE  ";
    my $r = $dbh->selectall_arrayref(
      "SELECT id, name ".
      "FROM account ".
      "WHERE id ". ($a ? '<' : '>') ." $MIN_PAYEE ".
      "ORDER BY name");
 
    if (!$compress) {
	print " $title                   TRANSACTIONS   BALANCE\n";
	print $LINE;
    }

    my $bal_sum = 0;
    my $trans_total = 0;
    foreach my $row (@{$r}) {
	my ($id,$name) = @{$row};

	my $trans = get_transactions($name);
	my $num = scalar @{$trans};
	my $lrow = $trans->[-1];
	$bal = $lrow ? $lrow->[-1] : 0;
	$bal_sum += $bal;
	$trans_total += $num;

	printf($ACC_FMT,$name,$num,$bal/100);
    }

    if (!$compress) {
	print $LINE;
	printf(" TOTALS                           %5d $MFMT\n",
	       $trans_total, $bal_sum/100);
	print $LINE;
    }

}

#----------------------------------------------------------
sub do_import
{
    my $file = $opts{f};
    open(FILE,$file) || die "Cannot open file $file";
    foreach (<FILE>) {
	chomp;
	if (/^\s*([0-9\-]+)\s+\"([^\"]+)\"\s+([0-9\-\.]+)\s*$/) {
	    ($date,$payee,$value) = ($1,$2,$3);
	    $type = 'c';
	    if ($value < 0) {
		$type = 'd';
		$value = -$value;
	    }
	    do_add();
	}
    }
    close(FILE);
}

#----------------------------------------------------------
sub get_transactions
{
    my ($account) = @_;
    my @trans = ();

    my $r = $dbh->selectall_arrayref(
      "SELECT t.id, date, a1.name AS 'from', a2.name AS 'to', value, comment ".
      "FROM transaction AS t, account AS a1, account AS a2 ".
      "WHERE t.acc_src = a1.id AND t.acc_dst = a2.id AND ".
      "(a1.name = \"$account\" OR a2.name = \"$account\") ".
      "ORDER BY date, t.id");

    my $bal = 0;
    foreach my $row (@{$r}) {
	my ($id,$date,$src,$dst,$value,$comment) = @{$row};
	my $payee = "";
	my $credit = 0;
	my $debit = 0;
	
	if ($src eq $account && $dst eq $account) {
	    $debit = $bal-$value if ($bal > $value);
	    $credit = $value-$bal if ($bal < $value);
	    $payee = '=BALANCE CHECKPOINT=';

	} elsif ($src eq $account) {
	    $debit = $value;
	    $payee = $dst;

	} elsif ($dst eq $account) {
	    $credit = $value;
	    $payee = $src;
	    
	} else {
	    die "Selected bad transaction";
	}
	
	$bal += $credit - $debit;
	my @row = ($id,$date,$payee,$comment,$debit,$credit,$bal);
	push(@trans,\@row);
    }
    
    my $tref = \@trans;
    return $tref;
}

#----------------------------------------------------------
sub lookup_account
{
    my ($name,$auto_create) = @_;

    my $r = $dbh->selectrow_arrayref(
      "SELECT id FROM account WHERE name = \"$name\"");
    my $id = $r->[0];

    if ($auto_create && !defined $id) {
	$r = $dbh->selectrow_arrayref(
          "SELECT MAX(id) FROM account");
	$id = 1 + $r->[0];
	print "Payee '$payee' not found...";
        $dbh->do(
           "INSERT INTO account (id, name) ".
           "VALUES ($id, \"$name\")");
	print " added (id=$id)\n";
    }

    return $id;
}

#----------------------------------------------------------
sub do_init_tables
{
    $dbh->do(
      "CREATE TABLE IF NOT EXISTS transaction ( ".
      "id       INT AUTO_INCREMENT PRIMARY KEY,".
      "date     DATE,".
      "acc_src  INT,".
      "acc_dst  INT,".
      "value    INT,".
      "comment  VARCHAR(255) )");

    $dbh->do(
      "CREATE TABLE IF NOT EXISTS account ( ".
      "id       INT UNIQUE PRIMARY KEY,".
      "name     VARCHAR(32) )");
    

    if (!lookup_account("MIN_PAYEE",0)) {
	$dbh->do("INSERT INTO account (id,name) VALUES (1000,'MIN_PAYEE')");

	# Create default accounts
	$dbh->do("INSERT INTO account (id,name) VALUES (0,'current')");
	$dbh->do("INSERT INTO account (id,name) VALUES (1,'credit card')");
	$dbh->do("INSERT INTO account (id,name) VALUES (2,'savings')");
    }
}

#----------------------------------------------------------
sub read_rcfile
{
    if (open(RC,$RCFILE)) {
	foreach (<RC>) {
	    chomp;
	    if (/^ *\#/) {
		# Comment
	    } elsif (/^ *$/) {
		# Blank line
	    } elsif (/^ *([^ ]+)\: *(.*) *$/) {
		$rcopts{$1} = $2;
	    } else {
		print "ERROR: Unknown line in $RCFILE: $_\n";
	    }
	}
	close(RC);
    }
}

#----------------------------------------------------------
sub do_help
{
print <<'END';
acow - Simple command-line accounting
usage: acow <command> {options}
  command is one of:
    add           Add a transaction
    delete        Delete a transaction
    edit          Edit a transaction
    copy          Copy a transaction
    list          List transactions on an account
    info          Display info about a transaction
    accounts      List my accounts with balances
    payees        List payee accounts with balances
    import        Import transactions from file
    help          Displays this useful help text
    version       Displays version and copyright
  options:
    -a <account>  Select account [default: current]
    -d <date>     Transaction date
    -t <type>     Transaction type:
      (d=debit [default], c=credit, b=balance checkpoint)
    -p <payee>    Payee account
    -v <value>    Transaction value
    -c <comment>  Comment on transaction
    -f <file>     File to import/export
    -i <id>       Transaction ID to edit/copy
    -e            Display extra information
END
}

#----------------------------------------------------------
sub do_ver
{
print "acow-$ACOW_VER (http://sconemad.com/acow)\n";
print "Copyright (c) 2006 Andrew Wedgbury <wedge\@sconemad.com>\n";
print "\n";
}
